x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/LogicalPlanOptimizer.java

```diff
@@ -603,6 +606,106 @@ protected LogicalPlan rule(OrderBy orderBy) {
+                    if (p instanceof Aggregate aggregate) {
+                        var pruned = seenProjection.get() ? removeUnused(aggregate.aggregates(), usedSet) : null;
+                        // no aggregates, no need
+                        if (pruned != null) {
+                            if (pruned.isEmpty()) {
+                                recheck = true;
+                                p = aggregate.child();
+                            } else {
+                                p = new Aggregate(aggregate.source(), aggregate.child(), aggregate.groupings(), pruned);
+                            }
+                        }
+
+                        seenProjection.set(Boolean.TRUE);
+                    } else if (p instanceof Eval eval) {
+                        var pruned = seenProjection.get() ? removeUnused(eval.fields(), usedSet) : null;
+                        // no fields, no eval
+                        if (pruned != null) {
+                            if (pruned.isEmpty()) {
+                                p = eval.child();
+                                recheck = true;
+                            } else {
+                                p = new Eval(eval.source(), eval.child(), pruned);
+                            }
+                        }
+                    } else if (p instanceof Project) {
+                        seenProjection.set(Boolean.TRUE);
+                    }
+                } while (recheck);
+
+                var inUse = usedSet.combine(references(p));
+                used.set(inUse);
+
+                // preserve the state before going to the next node
+                return p;
+            });
+
+            return pl;
+        }
+
+        /**
+         * Prunes attributes from the list not found in the given set.
+         * Returns null if no changed occurred.
+         */
+        private static <N extends NamedExpression> List<N> removeUnused(List<N> named, AttributeSet used) {
+            var clone = new ArrayList<>(named);
+            var it = clone.listIterator(clone.size());
+
+            boolean changed = false;
+            // due to Eval, go in reverse
+            while (it.hasPrevious()) {
+                N prev = it.previous();
+                boolean remove = used.contains(prev.toAttribute()) == false;
+                if (remove) {
+                    changed = true;
+                    it.remove();
+                } else {
+                    used = used.combine(prev.references());
+                }
+            }
+            return changed ? clone : null;
+        }
+
+        private static List<Expression> expressions(LogicalPlan plan) {
+            List<Expression> exp = new ArrayList<>();
+            plan.forEachExpression(exp::add);
+            return exp;
+        }
+
+        private static AttributeSet references(LogicalPlan plan) {
+            return Expressions.references(expressions(plan));
+        }
```

Comment:
question: would it make sense to make `references` a method of `LogicalPlan`? Maybe only makes sense once e.g. multiple optimzation rules need to access all the references of a plan, but I would expect this to be helpful convencience method (in addition to `forEachExpression`).

Reply:
Yes - in fact both expressions() and references() should be added to `QueryPlan` - waiting on the merge to main first.
https://github.com/elastic/elasticsearch/issues/98509